test/ranges.jl

```diff
@@ -1,5 +1,25 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
+@testset "range construction" begin
+    @testset "range(;kw...)" begin
+        @test_throws ArgumentError range(start=1, step=1, stop=2, length=10)
+        @test_throws ArgumentError range(start=1, step=1, stop=10, length=11)
+
+        r = 3.0:2:11
+        @test r == range(start=first(r), step=step(r), stop=last(r)                  )
+        @test r == range(start=first(r), step=step(r),               length=length(r))
+        @test r == range(start=first(r),               stop=last(r), length=length(r))
+        @test r == range(                step=step(r), stop=last(r), length=length(r))
+
+        r = 4:9
+        @test r === range(start=first(r), stop=last(r)                  )
+        @test r === range(start=first(r),               length=length(r))
+        # the next one uses ==, because it changes the eltype
```

Comment:
> My suggestion is that we use `===` to compare to similar `(:)`, colon, based syntax in situations where we expect them to perform similarly. Comparable output to `(:)` for similar input is a fairly minimal contract for this method.
> 
> ```julia
>         r = 4:9
>         @test range(start=first(r),               stop=last(r)                  ) === r
>         @test range(start=first(r),                             length=length(r)) === r
>         @test range(start=first(r),               stop=last(r), length=length(r))  == r # Does not produce a `UnitRange`
>         @test range(                              stop=last(r), length=length(r)) === r
> ```



@mkitti This is pretty close to your suggestion, looks good to you?

Reply:
Yes. I think you were already doing it. I was just pointing out that you are actually already checking for type `UnitRange` by using `===`.

A discussion for another day might be if `length == stop - start + 1` and both `start` and `stop` are `Integer`, then maybe we should produce a `UnitRange`. It might be as simple as adding a ternary operator.

```julia
range_start_stop_length(start::T, stop::T, len::Integer) where {T<:Integer} =
    len == stop - start + 1 ?
        range_start_stop(start, stop) : 
        _linspace(float(T), start, stop, len)
```

The type stability implications might be problematic there though since the above function could either return a `UnitRange` or a `StepRangeLen`